### PR TITLE
Refactor: Split conv backward to allow conditional gradient computation

### PR DESCRIPTION
**Pull request by AsherJingkongChen**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Not found.

### Changes

In crate `burn-tensor`, I split each of `conv[1-3]d_backward` and their **transposed** variants into three functions suffixed with `*_x_backward`, `*_weight_backward`, and `*_bias_backward`.

In crate `burn-autodiff`, I changed the backward implementations of conv ops to allow gradients being computed **if required**.

The public APIs appear to remain unchanged, since `Autodiff` backend (decorator) hides the changes.

### Testing

All tests are passed on Apple M2 Pro. I did not add any new test because the original tests should be adequate.

### CODE REVIEW

**Comment by AsherJingkongChen** [comment]
### Use case

This is M-SSIM loss function. I use this to compare the performance between the `main` and my PR. The changed version runs faster.

> `examples/mssim/src/main.rs`:

```rust
use burn::{
    backend::{Autodiff, Wgpu},
    tensor::{backend::Backend, Distribution, Tensor},
};
use burn_common::sync_type::SyncType;
use std::time::Instant;

fn main() -> anyhow::Result<()> {
    pretty_env_logger::formatted_builder()
        .parse_default_env()
        .try_init()?;

    // Settings
    type B = Autodiff<Wgpu>;
    let device = &Default::default();

    for _ in 0..5 {
        // Target image (No grad)
        let target = Tensor::<B, 4>::random([1, 3, 600, 900], Distribution::Default, device)
            .set_require_grad(false);

        // Rendered image (With grad)
        let rendered = target.clone().neg().add_scalar(1.0).set_require_grad(true);

        // Get MSSIM index
        let mssim = lib::MeanStructuralSimilarity::<B, 3>::init(device);
        let loss = mssim.forward(rendered.clone(), target.clone());

        // Backward pass
        let time = Instant::now();
        let grads = &mut loss.backward();
        B::sync(device, SyncType::Wait);

        // Duration records (fastest)
        // 6f0e61aa4f7368914c2492b11095a74145f4209c: 930.839125ms
        // 8d0852f60dba4eda4c05b9add7a57b48b0942f99: 41.750209ms
        println!("MSSIM backward pass duration: {:?}", time.elapsed());

        // Retrieving gradients
        let target_grad = target.grad_remove(grads);
        let rendered_grad = rendered.grad_remove(grads);
        let filter_weight_grad = mssim.filter.weight.grad_remove(grads);

        assert!(!target_grad.is_some());
        assert!(rendered_grad.is_some());
        assert!(!filter_weight_grad.is_some());
    }
    Ok(())
}

mod lib {
    use burn::{
        module::Param,
        nn::{conv, PaddingConfig2d},
        tensor::{backend::Backend, Int, Tensor},
    };

    /// Computing the mean of structural similarity index (MSSIM) between the inputs
    /// using the approaches described in the paper:
    ///
    /// *Wang, J., Bovik, A. C., Sheikh, H. R., & Simoncelli, E. P. (2004). Image quality assessment: from error visibility to structural similarity. IEEE Transactions on Image Processing, 13(4), 600–612.*
    /// https://www.cns.nyu.edu/pub/lcv/wang03-preprint.pdf
    ///
    /// ## Details
    ///
    /// * `self.filter.weight`: A normalized gaussian filter
    ///   with shape of `[C, 1, 11, 11]` and the standard deviation of `1.5`.
    ///   (No grad)
    #[derive(Clone, Debug)]
    pub struct MeanStructuralSimilarity<B: Backend, const C: usize> {
        pub filter: conv::Conv2d<B>,
    }

    impl<B: Backend, const C: usize> MeanStructuralSimilarity<B, C> {
        pub fn init(device: &B::Device) -> Self {
            const WEIGHT_SIZE: usize = 11;
            const WEIGHT_SIZE_HALF: usize = WEIGHT_SIZE >> 1;
            const WEIGHT_STD: f64 = 1.5;
            const WEIGHT_STD2: f64 = WEIGHT_STD * WEIGHT_STD;

            let mut filter = conv::Conv2dConfig::new([C; 2], [WEIGHT_SIZE; 2])
                .with_bias(false)
                .with_groups(C)
                .with_padding(PaddingConfig2d::Valid)
                .init(device);

            filter.weight = Param::uninitialized(
                Default::default(),
                move |device, is_required_grad| {
                    let size_half = WEIGHT_SIZE_HALF as i64;
                    let x = Tensor::<B, 1, Int>::arange(-size_half..size_half + 1, device);
                    let x2_n = x.powi_scalar(2).neg().float().unsqueeze::<2>();
                    let y2_n = x2_n.to_owned().transpose();
                    let x2_y2_n = x2_n + y2_n;
                    let w = x2_y2_n.div_scalar(2.0 * WEIGHT_STD2).exp();
                    let w_normalized = w.to_owned().div(w.sum().unsqueeze::<2>());

                    w_normalized
                        .expand([C, 1, WEIGHT_SIZE, WEIGHT_SIZE])
                        .set_require_grad(is_required_grad)
                },
                device.to_owned(),
                false,
            );

            Self { filter }
        }
    }

    impl<B: Backend, const C: usize> MeanStructuralSimilarity<B, C> {
        pub fn forward<const D: usize>(
            &self,
            value: Tensor<B, D>,
            target: Tensor<B, D>,
        ) -> Tensor<B, 1> {
            const K1: f64 = 0.01;
            const K2: f64 = 0.03;
            const L: f64 = 1.0;
            const C1: f64 = (K1 * L) * (K1 * L);
            const C2: f64 = (K2 * L) * (K2 * L);
            const FRAC_C1_2: f64 = C1 / 2.0;
            const FRAC_C2_2: f64 = C2 / 2.0;

            let input = (
                value.unsqueeze::<4>().expand([-1, C as i64, -1, -1]),
                target.unsqueeze::<4>().expand([-1, C as i64, -1, -1]),
            );

            debug_assert_eq!(input.0.dims(), input.1.dims());

            let filter = &self.filter;
            let mean = (
                filter.forward(input.0.to_owned()),
                filter.forward(input.1.to_owned()),
            );
            let mean2 = (
                mean.0.to_owned().powf_scalar(2.0),
                mean.1.to_owned().powf_scalar(2.0),
            );
            let std2 = (
                filter
                    .forward(input.0.to_owned().powf_scalar(2.0))
                    .sub(mean2.0.to_owned()),
                filter
                    .forward(input.1.to_owned().powf_scalar(2.0))
                    .sub(mean2.1.to_owned()),
            );
            let mean_01 = mean.0 * mean.1;
            let std_01 = filter.forward(input.0 * input.1).sub(mean_01.to_owned());
            let indexes = (mean_01 + FRAC_C1_2) * (std_01 + FRAC_C2_2) * 4.0
                / ((mean2.0 + mean2.1 + C1) * (std2.0 + std2.1 + C2));
            indexes.mean()
        }
    }
}
```

> `examples/mssim/Cargo.toml`:

```toml
[package]
edition = "2021"
name = "mssim"
version = "0.0.0"

[dependencies]
anyhow = {version = "1.0.86", default-features = false, features = ["std"]}
burn-common = {path = "../../crates/burn-common"}
log = {version = "0.4.22", default-features = false}
pretty_env_logger = {version = "0.5.0", default-features = false}

[dependencies.burn]
path = "../../crates/burn"
default-features = false
features = ["autodiff", "default", "wgpu"]
```

**Comment by nathanielsimard** [comment]
@

> Do we need the `WithBias` and `NoBias` variant of each convolution?

Because of the `prepare` method in the backward pass, it's not yet possible to have only one backward struct.

**Comment by AsherJingkongChen** [comment]
> @
> 
> > Do we need the `WithBias` and `NoBias` variant of each convolution?
> 
> Because of the `prepare` method in the backward pass, it's not yet possible to have only one backward struct.

@nathanielsimard I think so.
